// utils-nxg-cg/utils/constants.js
module.exports = {
    emits: {
        data: 'data',
        end: 'end',
        error: 'error',
        snapshot: 'snapshot'
    },
    messages_databases: {
        CONNECTION_END: 'Connection end',
        ERROR_BATCH: 'Batch size exceeds the limit of',
        QUERY_ERROR: 'Error executing query',
        QUERY_EXECUTED: 'Query successfully executed',
        ST: 'Start transaction',
        START_BATCH: 'Start batch query',
        START_COMMIT: 'Start commit',
        START_QUERY: 'Start single query',
        TRANSACTION_ERROR: 'Failed, undo changes and close connection'
    },
    ELASTICIO_LISTEN_MESSAGES_ON: process.env.ELASTICIO_LISTEN_MESSAGES_ON || null,
    md5sum: 'md5sum',
    ERROR_PROPERTY: 'Error missing property',
    ERROR_WITH_PROPERTY: 'Error with the property',
    FINISH_EXEC: 'Finished execution',
    PROCESS_SUCCESS: 'Process successfully completed',
    START_PROCESS: 'Start process',
    START_EXEC: 'Start execution',
    START_TRANS: 'Start transaction',
    SUCCESS_TRANS: 'Success transform',
    URI_RABBITMQ: process.env.URI_RABBITMQ || 'amqp://localhost:5672',
    ERROR_JSON_FORMAT:'Bad format JSON',
    ERROR_XML_FORMAT:'Bad format XML Structure',
    ERROR_MD5:'Failed to check md5 sum',
    ERROR_CONVERT_ENCODING : 'Error to valid, null, undefined or bad encoding to convert',
    log_levels: {
        info: "info",
        error: "error",
        debug: "debug",
        warn: "warn"
    },
    flags: {
        CREATEDIRECTORY: 'CREATEDIRECTORY',
        DELETEDIRECTORY: 'DELETEDIRECTORY',
        DELETEFILE: 'DELETEFILE',
        DOWNLOADIRECTORY: 'DOWNLOADIRECTORY',
        GETFILE: 'GETFILE',
        GETLISTFILES: 'GETLISTFILES',
        RENAMEFILE: 'RENAMEFILE',
        SAVEFILE: 'SAVEFILE',
        UPLOADIRECTORY: 'UPLOADIRECTORY'
    }
};
